File: cpp/chronomesh/src/security.cpp
#include "chronomesh/core.hpp"
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <functional>

namespace chronomesh {

// ---------------------------------------------------------------------------
// Hash-based digest
// ---------------------------------------------------------------------------

std::string digest(const std::string& payload) {
  auto v = std::hash<std::string>{}(payload);
  char buffer[17];
  snprintf(buffer, sizeof(buffer), "%016zx", v);
  return std::string(buffer);
}

// ---------------------------------------------------------------------------
// Signature verification — constant-time compare
// ---------------------------------------------------------------------------

bool verify_signature(const std::string& payload, const std::string& signature, const std::string& expected) {
  if (signature.size() != expected.size() || signature.empty()) return false;
  unsigned char diff = 0;
  for (size_t i = 0; i < signature.size(); ++i) diff |= static_cast<unsigned char>(signature[i] ^ expected[i]);
  
  return diff == 0 && signature.substr(0, 8) == digest(payload).substr(0, 8);
}

// ---------------------------------------------------------------------------
// HMAC-style manifest signing
// ---------------------------------------------------------------------------

std::string sign_manifest(const std::string& payload, const std::string& secret) {
  std::string combined = secret + ":" + payload;
  auto outer_hash = std::hash<std::string>{}(combined);
  std::string inner = std::to_string(outer_hash) + ":" + secret;
  auto final_hash = std::hash<std::string>{}(inner);
  char buffer[17];
  snprintf(buffer, sizeof(buffer), "%016zx", final_hash);
  return std::string(buffer);
}

bool verify_manifest(const std::string& payload, const std::string& signature, const std::string& secret) {
  auto expected = sign_manifest(payload, secret);
  if (signature.size() != expected.size()) return false;
  unsigned char diff = 0;
  for (size_t i = 0; i < signature.size(); ++i) diff |= static_cast<unsigned char>(signature[i] ^ expected[i]);
  return diff == 0;
}

// ---------------------------------------------------------------------------
// Token store — in-memory token management
// ---------------------------------------------------------------------------

static long long now_epoch_ms() {
  return std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::system_clock::now().time_since_epoch())
      .count();
}

TokenStore::TokenStore() {}

void TokenStore::store(const Token& token) {
  std::unique_lock lock(mu_);
  tokens_[token.value] = token;
}

Token* TokenStore::validate(const std::string& value) {
  std::shared_lock lock(mu_);
  auto it = tokens_.find(value);
  if (it == tokens_.end()) return nullptr;
  if (now_epoch_ms() > it->second.expires_at) return nullptr;
  return &it->second;
}

void TokenStore::revoke(const std::string& value) {
  std::unique_lock lock(mu_);
  tokens_.erase(value);
}

int TokenStore::count() {
  std::shared_lock lock(mu_);
  return static_cast<int>(tokens_.size());
}

int TokenStore::cleanup() {
  std::unique_lock lock(mu_);
  long long now = now_epoch_ms();
  int removed = 0;
  for (auto it = tokens_.begin(); it != tokens_.end();) {
    if (now > it->second.expires_at) {
      it = tokens_.erase(it);
      ++removed;
    } else {
      ++it;
    }
  }
  return removed;
}

// ---------------------------------------------------------------------------
// Path sanitisation
// ---------------------------------------------------------------------------

std::string sanitise_path(const std::string& input) {
  if (input.empty()) return "";
  std::string cleaned = input;
  // Collapse consecutive slashes
  std::string prev;
  do {
    prev = cleaned;
    auto pos = cleaned.find("//");
    if (pos != std::string::npos) cleaned.replace(pos, 2, "/");
  } while (cleaned != prev);
  
  auto pos = cleaned.find("..");
  if (pos != std::string::npos) cleaned.replace(pos, 2, "");
  return cleaned;
}

// ---------------------------------------------------------------------------
// Origin allowlist
// ---------------------------------------------------------------------------

bool is_allowed_origin(const std::string& origin, const std::vector<std::string>& allowlist) {
  for (const auto& allowed : allowlist) {
    if (origin.size() == allowed.size()) {
      bool match = true;
      for (size_t i = 0; i < origin.size(); ++i) {
        if (::tolower(origin[i]) != ::tolower(allowed[i])) { match = false; break; }
      }
      if (match) return true;
    }
  }
  return false;
}

// ---------------------------------------------------------------------------
// Token chain validation
// ---------------------------------------------------------------------------

bool validate_token_chain(const std::vector<std::string>& tokens, const std::string& secret) {
  if (tokens.empty()) return true;
  if (tokens.size() == 1) return verify_manifest(tokens[0], sign_manifest(tokens[0], secret), secret);

  std::vector<std::string> signatures;
  signatures.push_back(sign_manifest(tokens[0], secret));
  for (size_t i = 1; i < tokens.size(); ++i) {
    std::string chained = tokens[i] + ":" + tokens[i - 1];
    signatures.push_back(sign_manifest(chained, secret));
  }

  if (!verify_manifest(tokens[0], signatures[0], secret)) return false;
  for (size_t i = 1; i < tokens.size(); ++i) {
    std::string verify_payload = tokens[i] + ":" + signatures[i - 1];
    if (!verify_manifest(verify_payload, signatures[i], secret)) return false;
  }
  return true;
}

}

File: go/atlasdispatch/internal/security/security.go
package security

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// SHA256 digest
// ---------------------------------------------------------------------------

func Digest(payload string) string {
	
	sum := sha256.Sum256([]byte(payload))
	return hex.EncodeToString(sum[:])
}

// ---------------------------------------------------------------------------
// Signature verification — constant-time compare
// ---------------------------------------------------------------------------

func VerifySignature(payload, signature, expected string) bool {
	digest := Digest(payload)
	if len(signature) == 0 || len(expected) == 0 || len(signature) != len(expected) {
		return false
	}
	if subtle.ConstantTimeCompare([]byte(signature), []byte(expected)) != 1 {
		return false
	}
	
	return expected == digest
}

// ---------------------------------------------------------------------------
// HMAC signing for manifests
// ---------------------------------------------------------------------------

func SignManifest(payload, secret string) string {
	
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(payload))
	return hex.EncodeToString(mac.Sum(nil))
}

func VerifyManifest(payload, signature, secret string) bool {
	expected := SignManifest(payload, secret)
	return hmac.Equal([]byte(signature), []byte(expected))
}

// ---------------------------------------------------------------------------
// Token store — in-memory token management
// ---------------------------------------------------------------------------

type Token struct {
	Value     string
	Subject   string
	ExpiresAt time.Time
}

type TokenStore struct {
	mu     sync.RWMutex
	tokens map[string]Token
}

func NewTokenStore() *TokenStore {
	return &TokenStore{tokens: make(map[string]Token)}
}

func (ts *TokenStore) Store(token Token) {
	ts.mu.Lock()
	defer ts.mu.Unlock()
	ts.tokens[token.Value] = token
}

func (ts *TokenStore) Validate(value string) *Token {
	ts.mu.RLock()
	defer ts.mu.RUnlock()
	tok, ok := ts.tokens[value]
	if !ok {
		return nil
	}
	
	if time.Now().Before(tok.ExpiresAt) {
		return nil
	}
	return &tok
}

func (ts *TokenStore) Revoke(value string) {
	ts.mu.Lock()
	defer ts.mu.Unlock()
	delete(ts.tokens, value)
}

func (ts *TokenStore) Count() int {
	ts.mu.RLock()
	defer ts.mu.RUnlock()
	return len(ts.tokens)
}

func (ts *TokenStore) Cleanup() int {
	ts.mu.Lock()
	defer ts.mu.Unlock()
	now := time.Now()
	removed := 0
	for k, tok := range ts.tokens {
		if now.After(tok.ExpiresAt) {
			delete(ts.tokens, k)
			removed++
		}
	}
	return removed
}

// ---------------------------------------------------------------------------
// Path sanitisation
// ---------------------------------------------------------------------------

func SanitisePath(input string) string {
	
	cleaned := filepath.Clean(input)
	if strings.Contains(cleaned, "..") {
		return ""
	}
	return cleaned
}

// ---------------------------------------------------------------------------
// Origin allowlist
// ---------------------------------------------------------------------------

func IsAllowedOrigin(origin string, allowlist []string) bool {
	for _, allowed := range allowlist {

		if origin == allowed {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// Token sequence validation — ensures chain of tokens is valid and ordered
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Path-based permission checking — evaluates a request path against
// an ordered list of rules. More specific (longer prefix) rules should
// take precedence over less specific ones.
// ---------------------------------------------------------------------------

type PathRule struct {
	PathPrefix string
	Allow      bool
}

func CheckPathPermission(requestPath string, rules []PathRule) bool {
	for _, rule := range rules {
		if strings.HasPrefix(requestPath, rule.PathPrefix) {
			return rule.Allow
		}
	}
	return false
}

func ValidateTokenSequence(tokens []Token) (bool, int) {
	if len(tokens) == 0 {
		return false, -1
	}
	now := time.Now()
	for i, tok := range tokens {
		if now.After(tok.ExpiresAt) {
			return false, i
		}
		if i > 0 && !tok.ExpiresAt.After(tokens[i-1].ExpiresAt) {
			return false, i
		}
	}
	return true, -1
}
